### PR TITLE
Refactor coingecko

### PR DESCRIPTION
**Pull request by ls13g12**
Created /types with types for coingecko data
Refactored and moved coingecko helper functions to /utils
Setup jest with tests for coingecko helper, including mock api response from coingecko
Added messages above search bar, to prompt user to refine search.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'node',
+};
```
Comment (adawx) [inline jest.config.js]: nit: no newline at eof

File: pages/search.tsx [hunk 2]
```diff
@@ -1,24 +1,41 @@
...
+import { Coin } from '../types/Coin';
 
 const Search = () => {
-    const [newCoin, setNewCoin] = useState('')
-    const [coins, setCoins] = useState([])
-
-    const handleNewCoin = async (event: any) => {
-
+    const [newCoin, setNewCoin] = useState<string>('')
+    const [coins, setCoins] = useState<Coin[]>()
```
Comment (adawx) [inline pages/search.tsx]:
```suggestion
    const [coins, setCoins] = useState<Coin[]>([])
```

Should generally initialise as an empty array.

File: pages/search.tsx [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
+
+                //only show 5 or fewer coins  - currently hardcoded
+                if (filteredCoins.length <= 5){
+                    setCoins(filteredCoins)
+                    setMessage('Here are your coins!')
+                }
+                else {
+                    setMessage('Refine your search')
+                    setCoins([])
+                }
```
Comment (adawx) [inline pages/search.tsx]: Could just show the top 5 regardless & optionally show a message showing that there are further results.

File: pages/search.tsx [hunk 4]
```diff
@@ -28,18 +45,13 @@ const Search = () => {
             </form>
             <div>
                 <ul>
-                {coins.map((coin: Coin) => (
-                    <li key={coin.id}>{coin.id}</li>
+                {coins?.map((coin: Coin) => (
+                    <li key={coin.id}>{coin.id} - rank: {coin.market_cap_rank}</li>
                 ))}
                 </ul>
             </div>
```
Comment (adawx) [inline pages/search.tsx:41]: You could extract this out to a `SearchResults` component and pass the coin list in as props and conditionally render based on coin list.

File: utils/coingecko_helper.test.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  ]
+}
+
+
+global.fetch = jest.fn(() => Promise.resolve({
+    json: () => Promise.resolve(MOCK_DATA)
+})) as jest.Mock;
+
+
+describe("test coingecko api search", () => {
```
Comment (adawx) [inline utils/coingecko_helper.test.ts:47]: Nice! Once this is merged then `npm test` can be added into github actions to run on PR's

File: pages/search.tsx (hunk 4 above)
Comment (ls13g12) [reply pages/search.tsx:41]: I'll do that when I setup adding coins to portfolio functionality.

**Pull request merged**